pilot/pkg/config/coredatamodel/discovery.go

```diff
@@ -0,0 +1,331 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package coredatamodel
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+
+	networking "istio.io/api/networking/v1alpha3"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pkg/config/host"
+	"istio.io/istio/pkg/config/labels"
+	"istio.io/istio/pkg/config/protocol"
+	"istio.io/pkg/log"
+)
+
+var (
+	notReadyEndpointkey                        = "networking.alpha.istio.io/notReadyEndpoints"
+	_                   model.Controller       = &MCPDiscovery{}
+	_                   model.ServiceDiscovery = &MCPDiscovery{}
+)
+
+// DiscoveryOptions stores the configurable attributes of a Control
+type DiscoveryOptions struct {
+	XDSUpdater   model.XDSUpdater
+	Env          *model.Environment
+	ClusterID    string
+	DomainSuffix string
+}
+
+// mixerEnabled checks to see if mixer is enabled in the environment
+// so we can set the UID on eds endpoints
+func (o *DiscoveryOptions) mixerEnabled() bool {
+	return o.Env != nil && o.Env.Mesh != nil && (o.Env.Mesh.MixerCheckServer != "" || o.Env.Mesh.MixerReportServer != "")
+}
+
+// MCPDiscovery provides storage for NotReadyEndpoints
+type MCPDiscovery struct {
+	// [ip:port]config
+	notReadyEndpoints   map[string]*model.Config
+	notReadyEndpointsMu sync.RWMutex
+	options             *DiscoveryOptions
+}
+
+// NewMCPDiscovery provides a new instance of Discovery
+func NewMCPDiscovery(options *DiscoveryOptions) *MCPDiscovery {
+	return &MCPDiscovery{
+		notReadyEndpoints: make(map[string]*model.Config),
+		options:           options,
+	}
+}
+
+// RegisterNotReadyEndpoints registers newly received NotReadyEndpoints
+// via MCP annotations
+func (d *MCPDiscovery) RegisterNotReadyEndpoints(conf *model.Config) {
+	d.notReadyEndpointsMu.Lock()
+	defer d.notReadyEndpointsMu.Unlock()
+
+	if nrEps, ok := conf.Annotations[notReadyEndpointkey]; ok {
+		addrs := strings.Split(nrEps, ",")
+		for _, addr := range addrs {
+			d.notReadyEndpoints[addr] = conf
+		}
+	}
+}
+
+// GetProxyServiceInstances returns service instances co-located with a given proxy
+func (d *MCPDiscovery) GetProxyServiceInstances(proxy *model.Proxy) ([]*model.ServiceInstance, error) {
+	out := make([]*model.ServiceInstance, 0)
+
+	// There is only one IP for kube registry
+	proxyIP := proxy.IPAddresses[0]
+
+	d.notReadyEndpointsMu.Lock()
+	defer d.notReadyEndpointsMu.Unlock()
+	for addr, conf := range d.notReadyEndpoints {
```

Comment:
As from the above `RegisterNotReadyEndpoints`, the multiple not ready eps addressed can be belong to the same `conf`(service entry), [below]https://github.com/istio/istio/pull/12276/files#diff-eec332eccc20a758b8bb1e90490c3dc3R108) you calc service instances for the same SE multi times. So this is not right.